org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/ClasspathJsr199.java

```diff
@@ -250,6 +294,22 @@ public boolean hasAnnotationFileFor(String qualifiedTypeName) {
 	public Collection<String> getModuleNames(Collection<String> limitModules) {
 		if (this.jrt != null)
 			return this.jrt.getModuleNames(limitModules);
+		if (this.location.isModuleOrientedLocation()) {
+			Set<String> moduleNames = new HashSet<>();
+			try {
+				for (Set<Location> locs : this.fileManager.listLocationsForModules(this.location)) {
+					for (Location loc : locs) {
+						String moduleName = this.fileManager.inferModuleName(loc);
+						if (moduleName != null)
+							moduleNames.add(moduleName);
+					}
+				}
+				return moduleNames;
```

Comment:
If the implementation is good, it could perhaps benefit from caching.

But even the sequence of invocations looks funny: we list module locations, so those locations are already known to contain a module. Is it really necessary to then "infer" the module name? Wait, it seems we could use `if (loc instanceof ModuleLocationWrapper wrapper) .. wrapper.getModuleName()`.

To do so we'd need to
* make that class public
* provide a `getModuleName()` accessor

Next, I'm not even sure if it is safe to use our classes like LocationWrapper. Wouldn't that break if a custom file manager is used??

Reply:
> Wait, it seems we could use ... ModuleLocationWrapper ...

Never mind, `inferModuleName()` is implemented in that exact way I was considering for the client side, so this part looks OK :)

Still caching the set of module names might be a good idea.